src/vellum/utils/templating/render.py

```diff
@@ -14,6 +14,25 @@ def finalize(obj: Any) -> str:
     return str(obj)
 
 
+def custom_replace(s, old, new):
+    """
+    Custom replace filter that uses DefaultStateEncoder for conversion
+    """
+
+    def encode(obj):
```

Comment:
[blocking] I think `str` returns `None` instead of itself

I'm surprised that mypy didn't ask for types here. I'd be a fan of this method being pulled out and adding type hints to both